### PR TITLE
Scripts/ScarletMonastery: Rewrite Mograine and Whitemane encounter

### LINKED ISSUE #15793: Scarlet Monastery/Cathedral: High Inquisitor Whitemane's door is open
TrinityCore rev. 60da7c655273 2015-10-31 16:04:14 +0100 (HEAD branch)

The door must be closed

kloster

### PR DESCRIPTION
**Pull request by Artamedes**
**Changes proposed:**

-  Rewrite the Mograine and Whitemane encounter in Scarlet Monastery
-  also fix the door issue

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #15793 

**Tests performed:** (Does it build, tested in-game, etc.)
Builds and tested ingame

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp [hunk 1]
```diff
@@ -61,321 +53,391 @@ enum Spells
...
+        {
+            m_CanTalkKill = true;
+            m_CanDie = false;
+            m_FakeDeath = false;
+        }
 
         void Reset() override
         {
-            Initialize();
+            m_CanTalkKill = true;
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp]: keeping the old Initialize() way reduce code duplication, so these variables are set to their default value only in a function

**Comment by jackpoz** [comment]
> * Rewrite the Mograine and Whitemane encounter in Scarlet Monastery

what did the rewrite change ? what did it fix (other than the door issue) ?

File: src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp [hunk 2]
```diff
@@ -61,321 +53,391 @@ enum Spells
...
+            events.ScheduleEvent(EVENT_HAMMER_OF_JUSTICE, Seconds(10), Seconds(15));
         }
 
-        void KilledUnit(Unit* /*victim*/) override
+        void KilledUnit(Unit* who) override
         {
+            if (who->GetTypeId() != TYPEID_PLAYER || !m_CanTalkKill)
+                return;
+
+            // Prevent talking more than once - resets every 5 seconds by EVENT_RESET_KILL_TALK
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp]: what about using a single time_t variable instead ? would that look less complicated than a flag and an event resetting it ?

**Comment by Artamedes** [comment]
> > * Rewrite the Mograine and Whitemane encounter in Scarlet Monastery
> 
> what did the rewrite change ? what did it fix (other than the door issue) ? what was the bug with the door ?

The rewrite changed Scarlet Commander Mograine's AI (3976) and High Inquisitor Whitemane's AI (3977). The reason for rewrite is because I took a look at the script, and noticed it was using old codestyle, and is probably an old SD2 script.. So I decided to rewrite it, using new script styles, while documenting a bit, and still keeping mostly old script logic, but changing a few things to make it blizzlike.

For example: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp#L324 should've been AOE cast, and she should have also moved near mograine's body and faced him, then started resurrection.

This rewrite probably fixes all issues with this boss, however if some still exist, it can be easily modified and fixed

so the door issue was the door was always open for this encounter, however the door should only open, and stays open after the Mograine first dies (he's later resurrected in the encounter)

**Comment by Aokromes** [comment]
tested on http://tcubuntu.northeurope.cloudapp.azure.com/ works as intended.

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**